scala3/getting-started.md

```diff
@@ -0,0 +1,224 @@
+
+
+### If you don’t have those tools installed
+
+If you don’t already have Java and sbt installed on your system, and you’d like a simple way to install them — along with other popular tools in the Scala ecosystem — a great approach is to use a tool named [Coursier](https://get-coursier.io/docs/cli-overview).
+
+Coursier is an “artifact-fetching” tool written in Scala. It’s designed to automatically fetch dependencies for you and works in parallel, so the artifacts you need are downloaded as rapidly as possible.
+
+### Installing Coursier
+
+The <a href="https://get-coursier.io/docs/cli-installation" target="_blank">Coursier installation page</a> shows how to install Coursier and its `cs` command on macOS, Windows, and Linux systems. See that page for how to install Coursier, and then come back to this page.
+
+
+### Run the Coursier setup command
+
+<!-- TODO: need to be more clear about what JDK is installed -->
+With Coursier’s `cs` command-line tool installed, run this `cs setup` command to install a JDK, sbt, and the other tools mentioned above:
+
+```sh
+$ cs setup
+```
+<!-- TODO: show the Coursier output here -->
+
+>If you prefer more control on what `cs` does, see [its “setup” command page](https://get-coursier.io/docs/cli-setup) for more configuration options.
+
+After that command runs, you can verify that the tools you need are installed on your system and in your `PATH` with these commands:
+
+```sh
+$ java -version
+$ sbt --script-version
+```
+
+The output of those commands should show that you now have Java 11 and sbt 1.4.4 or newer installed. Assuming that worked, you’re now ready to create a “Hello, world” project with Scala 3 and sbt.
+<!-- TODO: verify that it installs Java 11 -->
+
+
+### What Coursier installs
+
+<!-- TODO: be more clear about “if it’s not already installed” -->
+That command installs all of the following software, if it’s not already installed:
+
+- A JDK
+- The [sbt](https://www.scala-sbt.org) and [mill](https://www.lihaoyi.com/mill) build tools
+- [Ammonite](https://ammonite.io), an enhanced REPL
+- [scalafmt](https://scalameta.org/scalafmt), the Scala formatter
+- The [Coursier CLI](https://get-coursier.io/docs/cli-overview), to install further Scala-based applications
+- The `scala` and `scalac` command-line tools
+<!-- TODO: be more clear about what JDK is installed -->
+
+Later, when you want to update your tools, use this command to update the installation:
+
+```sh
+cs update
+```
+
+## <a name="hello-world-example"></a>Create a “Hello, world” example with sbt
+
+Like Ant, Maven, and Gradle, sbt is a build tool that you use to build real-world applications. Like Maven, sbt uses “convention over configuration,” so you can create a “Hello, world” project in just a few steps. We’ll do that manually at first so you can see how sbt works, and then show a more automated way to do a similar thing.
+
+### Create an sbt project
```

Comment:
@alvinj is probably the best one to answer this. I think he wanted to show the things necessary to setup a project. However, for getting started `sbt new` might be better. Which template should be used to generate the proper Scala 3 project?

Reply:
we can use `sbt new scala/scala3.g8`

Reply:
I like `sbt new`, but I’m concerned that it has some issues that can be a turnoff to a first-time Scala developer. I put these notes in the *scala-tools.md* file in the book:

Using something like `sbt new scala/scala3.g8` may eventually be preferable, but that seems to have a few bugs atm:

- Creates *target* and *project* directories in the directory where you run `sbt new`
- Creates a directory with a lowercase version of the project name I supply (when it prompts me, I use a name like *HelloWorld*, and it creates a directory named *helloworld*)
- The *build.sbt* `name` doesn’t match the project name
- The *build.sbt* syntax may be a little hard for beginners

Those are the reasons I don’t jump right into using `sbt new`.

Also, it’s my mistake that I didn’t report those in the sbt project repository. I was just working on these docs and thought, “Hmm, I don’t like the way that works, let’s start with something simpler.” I can report these as issues if you agree that they’re a problem.

Reply:
`sbt new` actually uses https://github.com/foundweekends/giter8 underneath and you can run it via coursier. That's what we are doing in Metals. It would not create the additional directories. 

Also, `build.sbt`'s name can be changed in the template to correspond to the name that's given to it, so that should be easily fixable in the template.  Creating a directory with a lowercase is not exactly a bug I would say and is it a problem really?

We can make an additional beginner friendly template to solve any other issues. I think it would be nice to show how easy it is to create a new project.

Reply:
Yes, I submitted a PR on that template a few days ago to change all “Dotty” references to “Scala 3”. I have it on my list to see if the `name` could be changed, but I didn’t look into how to do that yet.

The directory name may just be a personal thing, but if I create a project named `FooBarBaz` and it’s renamed to `foobarbaz`, the first thing I do is rename that directory back to my name before `cd`ing into it (along with deleting the *project* and *target* directories). For things like this, I just want it to be a flawless first impression.

I forgot about the “Hello, world” template, and it has a nice *build.sbt* file for beginners. I’d be in favor of using an updated version of that one here, and then mentioning that they can use the *scala3.g8* template once they’re comfortable.

You mention Coursier, is there a `cs` command that can do this, and if so, could/would that be preferable? We’re already asking users to install Coursier, so I think they could use either. I don’t see a command on the Coursier site.

Also, if I’m making too big of a deal about these issues, just let me know.

Reply:
You can do for example:
```
cs install giter8
g8 playframework/play-scala-seed.g8 --name=MyProjectName
```

Reply:
The name can be changed I think via:
```
/build.sbt
name := "$name$"
```

Reply:
Thanks! I’ll try to look into this tonight (after I’ve finished my day job). ;)

Reply:
I think this is resolved, so I’m resolving this conversation. I’m also making a note on my personal to-do list to update that template, if nobody else has done that already. Thanks!